### ISSUE TITLE
Refactor client-server architecture

### ISSUE THREAD

**Issue by indomitableSwan** [boltlabs-inc/key-mgmt#98]
## Context 
We are looking at implementing our client-server architecture for the Lock Keeper (DAMS) project using grpc, specifically with the Rust [tonic](https://docs.rs/tonic/latest/tonic/) library. To date, we have been using the in-house [dialectic](https://docs.rs/dialectic/latest/dialectic/) library, but this library does not already have all the [features we need](https://github.com/boltlabs-inc/key-mgmt-spec/blob/main/systems-architecture.md#networking), and extending `dialectic` to our current and future use cases (e.g., handling multi-party protocols) is technically fraught. To lower our technical risk and ease future development, we would like to refactor to use the much more standard and well-supported `tonic` library instead.

However, the `tonic` library also does not provide all of the properties we want out of the box. With this in mind, we need a technical assessment and recommendation for how best to move forward on session handling and protocol implementations. This involves investigating how best to achieve the following properties, making a recommendation to the team on this thread, and propagating the outcomes of the ensuing discussion to development issues as appropriate. These properties are treated more-or-less in priority order.

This epic is related to #30, which may provide some further context as to how we have been thinking about sessions, but may need to be superseded by appropriately scoped issues in this epic.

## Expected Outcome
The overarching goal of this epic is to define our desired client-server architecture and the steps needed to refactor, including:
- [x] A detailed assessment and recommendation as to how best to realize our logical sessions, wherein messages are sent over a secure channel and the sessions satisfy the desired properties as defined below. 
- [x] Guidance on how to handle our protocol implementations moving forward. This includes:
    - [x] An implementation plan for the as-yet-unimplemented protocols. 
    - [x] A plan for refactoring our existing protocol implementations (for "register" and "authenticate" flows) using our new methodology, as appropriate.

An important caveat for the creation of this recommendation and guidance is that we do not have to implement "complete" sessions in our current development phase, i.e., it is desirable for us to move forward on protocol implementations quickly and to that end, we can defer realization of (some subset of) our desired session properties to later. We do not, however, wish to implement our protocols now in a way that makes realizing our desired sessions more difficult later. For example, the first section below details requirements for an additional communications encryption layer, and there is an immediate choice to be made that directly impacts how we represent protocol messages.

### Sessions that use TLS and an extra encryption layer 
After a client has already registered with the system, clients make requests to the server in a _Request session_, as described [here](https://github.com/boltlabs-inc/key-mgmt-spec/blob/main/systems-architecture.md#request-session). In broad strokes, connections between the client and server will take the form of an initial TLS handshake, followed by an OPAQUE handshake. Once the OPAQUE handshake is completed, all future session messages should be encrypted under a key derived from `session_key`, which is a shared key output to both the client and server from the OPAQUE handshake.

The `tonic` library has built-in support for TLS, but does not have a mechanism by which to nest the additional layer of communications encryption. Two possible ways of achieving this additional layer are:

1. Use the existing `tonic` library as is. If we go this route, this might mean that the protobuf definitions are not particularly meaningful as an interface definition language. That is, the messages that go back and forth are just byte streams that are encrypted, and the message structures themselves live in Rust, before encryption. The minimal solution here would seem to necessitate an efficient way to:
    1. Do an rpc call over TCP using TLS.
    1. Do the OPAQUE handshake.
    1. Stream encrypted bytes between the endpoints.
1. Build a customer encryption layer. This approach may allow us to incorporate our additional encryption layer in the same way that TLS is incorporated into `tonic` already. The advantage of this approach is that we may be able to write messages as normal protobufs. The disadvantage is that this seems like more engineering effort up front. 

It seems likely that using the existing `tonic` library is a safer choice here, and that a later refactor to add a custom layer to grpc would not be problematic, should we decide to do this in the future. However, we would benefit from a more nuanced technical assessment on this question, and whatever we decide, we need a concrete plan for implementation.y

Issue #30 has some notes about this step in the more generic, multi-party case. It might be useful context, but it's not appropriately scoped for this phase.

### Custom TLS configurations
We will be securing connections over TLS. We will need to:
- Have a custom certificate issuer. The client should trust _only_ our custom cert issuer.
- Require minimal TLS v1.3.
- Allow only our selected ciphersuite to be used.
- Have a plan for incrementally upgrading the network when a new version of TLS comes out.

### Session resumption
It is unrealistic to assume that the underlying connections never fail, particularly since one of our major use cases is a client application that runs on (and therefore makes requests from) an end user's device. Therefore it is desirable to support session resumption between different actual underlying connections.

### Multiplexing connections
In the near future, we will be supporting server-to-server connections secured over TLS and we would like to have the option of multiplexing logical sessions over a single underlying connection. This fits the use case of supporting automated flows. That is, we will have:
1. A `remote-client` that is integrated into a host, where this host is run by the service provider.
1. The `remote-client` connects to our `key-server` over a TLS connection with 2-way custom certs, each with a custom configuration as described above.
1. The `remote-client` will make requests on behalf of end users in an automated fashion, and there may be many logical sessions at any given point in time. We want these requests to have efficient underlying networking, which likely means not opening up unique underlying connections for each request.

### SGX enclave integration
We intend to integrate a secure processor (likely SGX enclaves) into our `key-server` component. At a minimum, the additional encryption layer must be handled by the enclave, not the host server. Our threat model dictates that the less information revealed to the host server, the better. This leaves an open question of how to appropriately handle client-server connections once we attempt to integrate SGX, and how this interacts with our usage of `tonic` (if at all).

**Comment by jakinyele** [comment]
Quick note from chat with @indomitableSwan about this part of the epic. We should focus on the first approach using the `tonic` library as is to implement the reg/auth protocol.

> 1. Use the existing `tonic` library as is. If we go this route, this might mean that the protobuf definitions are not particularly meaningful as an interface definition language. That is, the messages that go back and forth are just byte streams that are encrypted, and the message structures themselves live in Rust, before encryption. The minimal solution here would seem to necessitate an efficient way to:
>     1. Do an rpc call over TCP using TLS.
>     1. Do the OPAQUE handshake.
>     1. Stream encrypted bytes between the endpoints.
> 

We can investigate the custom encryption layer on top of protobufs as an optimization step later.

**Comment by picklenerd** [comment]
# Multiplexing connections
This functionality is built into `tonic`. There's a nice example in the `tonic` repo to show how it's done: https://github.com/hyperium/tonic/blob/master/examples/src/multiplex/client.rs


# Custom TLS configurations
`rustls` has robust configuration options that will let us specify [TLS version](https://docs.rs/rustls/latest/rustls/struct.ConfigBuilder.html#method.with_protocol_versions), [allowed ciphersuites](https://docs.rs/rustls/latest/rustls/struct.ConfigBuilder.html#method.with_cipher_suites), and [how certificates are verified](https://docs.rs/rustls/latest/rustls/client/trait.ServerCertVerifier.html).

See [ConfigBuilder](https://docs.rs/rustls/latest/rustls/struct.ConfigBuilder.html) for all configuration options.


# Session resumption
We will need to figure out things like where to store session state (on the server, on something like redis), how long to store incomplete sessions, and if there's any kind of attack surface that opens up based on how we implement this.

Session resumption is going to be a complicated problem that I think we can tackle after the PoC is finished.

**Comment by picklenerd** [comment]
## Protocol implementation

TLS will be handled by `tonic` but the extra layer of encryption will need to happen within the handlers we write for each of the endpoints.

Vidya and I explored how we can best implement this encrypted message-passing scheme in `tonic` in the [dams-tonic-toy repo](https://github.com/boltlabs-inc/dams-tonic-toy/tree/message-passing-experiment). It works as follows

1. Use `tonic` to establish a connection and route the request to the correct endpoint (e.g. registration)
2. Open channels between the client and server to pass messages back and forth
3. Messages are processed in their expected order by the client or server
4. If a message is received out of order, an error is returned and the connection is closed
5. Once the entire protocol is processed, the connection is closed

I think this is a good model to get us started but it's a bit too cumbersome and prone to developer error to be a long-term solution.
We have some convenient helper functions that keep the protocol implementation logic simple and linear but there will be a lot of repetition between those helpers and it will likely take some time to generalize them.

We plan to implement the initial PoC using this scheme and will design a better abstraction as soon as we have a working PoC.
Ideally this should happen before the PoC deadline. If it's not, this should be the top priority before we move forward with the next phase to minimize refactoring.

The advantage to this approach is that it should give us a working PoC sooner and give us time to properly design a system that works with all of the protocols we need to implement in the future.
We'll have to do a small amount of refactoring on the existing endpoints to update them to this new system, but the existing endpoints will help inform our design desicions give us working examples
of how this system consumed by code with real protocols.

We can discuss the details of how we implement this later, but my initial thought is to use a generic message type for all of the messages passed between the server and client. The protobuf definition would look something like this:

```
message EncryptedMessage {
  // Encrypted bytes
  bytes content = 1;
}
```

The handler would receive the message, decrypt the bytes, then deserialize the message into the next type it expects as defined by the protocol. Protocols could then be implemented using normal Rust types instead clunky generated types.
I have some ideas for how this might be implemented but we should table this for now and get a working PoC first.

**Comment by picklenerd** [comment]
## Sessions that use TLS and an extra encryption layer

I investigated implementing the extra encryption layer as a `tower` middleware but I don't think this will work.
The channels that we establish are part of the schema definition that `tonic` uses which means that the messages passed through those channels are processed in the handler functions that are part of trait generated by `tonic`.
Middlewares can add steps before and after the handler runs but they can't add steps in the middle of the handler.

If we use the generic message format I described above, we should be able to implement this encryption layer as part of the message processing system we develop after the PoC. It would work something like this:

1. Pull the next message from the channel
2. Decrypt the bytes stored in the `content` field
3. Deserialize the decrypted bytes into the expected message type
4. Process the message
5. If there is a response, encrypt it and create a new `EncryptedMessage`
6. Send the response over the channel

The developer writing the protocol should only have to work with the relevant Rust types to implement the protocol and should never have to worry about encryption or serialization.

**Comment by picklenerd** [comment]
The design for the new `tonic` service is spread across a few other issues.

## Server
The server architecture is a simple tonic service with a few boilerplate reduction functions. It accepts client connections, creates a channel to the client, and executes protocols as defined by the spec.

## Client
#175 captures the requirements of the client. The initial client was created in #172 but we still need to implement things like logical sessions and session resumption, which are tracked in #175.

The initial client has the following responsibilities:

- Store state that is used between requests
- Store common infrastructure such as RNGs
- Store a DamsRpcClient that can be used to create channels for each request type
- Create a channel abstraction that can be used in protocol implementation
- Execute protocols as defined by the spec

## Messages
#128 captures the design of the system that handles message passing across channels. 

- Messages represent data that needs to be transmitted for each step of a protocol
- Messages are transmitted as generic blobs of bytes
- Messages types are defined in the `dams::types` module
- The channel abstraction converts incoming and outgoing messages to the correct message type

## Desired Properties
- TLS: #198 implements TLS with the option to specify custom TLS configurations
- Session resumption: Tracked in #175 and #213
- Multiplexing connections: `tonic` has this baked in
- SGX: We don't have any tickets for this but First Foundry is working on it
